### ISSUE TITLE
C# version (Triangle.Net)

### ISSUE THREAD

**Issue by epsi1on** [wo80/Triangle#44]
Hi,
I think you did ported this project to C#, but there is no such project under your account. Means you do not maintain the `Triangle.Net` project anymore? I have some issues with it :) May i ask it here?

**Comment by wo80** [comment]
Feel free to ask here.

I'm watching the fork at https://github.com/Geri-Borbas/Triangle.NET, so I'll get noticed about issues posted there. Also be aware of the [licensing issues](https://github.com/Geri-Borbas/Triangle.NET/issues/6), so I'd discourage you from using the code in a commercial context or creating a hard dependency on it in your open source projects.

**Comment by epsi1on** [comment]
I have these:


- An outer ring as a list of points (black one)
- Several holes, also they are in form of points (red ones)
- Several line segments, which should be used for constraints (green one). by constaint i mean no triangle should cross this constraint

![inputs](https://user-images.githubusercontent.com/7839141/153595635-fea99a52-f4e7-42c3-be90-fdd8a0295f1d.png)


The red holes can have collisions with the outer ring also green lines. 
Does the package handle this case by itself? Or I should do preprocess my geometry data?
Thanks

**Comment by wo80** [comment]
I don't think there's a straight forward way to handle such a geometry. To get started, here's a gist:
https://gist.github.com/wo80/fe9b40925cdfdafedad935457c4768dd

You can then use the code from this example to process only the triangles that are of interest:
https://github.com/garykac/triangle.net/blob/master/wiki/Example_7.md

**Comment by wo80** [comment]
And always be aware that Triangle.NET shares the same restrictions as the original C code:
http://www.cs.cmu.edu/~quake/triangle.trouble.html

**Comment by epsi1on** [comment]
Thanks.
Do you accept donations on your projects? I'm interested to donate crypto or some local handicrafts... :)

**Comment by epsi1on** [comment]
Is it a suitable input geometry for triangulation?
![bitmap2](https://user-images.githubusercontent.com/7839141/153697600-1c60cdd1-3df2-4e22-8e0a-f36337929bb1.png)

**Comment by wo80** [comment]
> Is it a suitable input geometry for triangulation?

Yes, that's a much saner input, but might require a segment intersection algorithm, which is `O(n log n)`, just like Delaunay triangulation.

I've updated the gist with a little mesh post-processing function, which returns
```
Processed 40 triangles in region connected to segment with label 1
Processed 0 triangles in region connected to segment with label 3
Processed 10 triangles in region connected to segment with label 3
Processed 0 triangles in region connected to segment with label 4
Processed 39 triangles in region connected to segment with label 4
```
Label 1 is the outer boundary, label 4 the constraint. So the only problem is, that the hole (label 3, purple region in image below) is also processed. This is because the outer boundary intersects the hole, which defeats the automatic labeling of regions.

![issue-44-2](https://user-images.githubusercontent.com/1840402/153712715-a04119f6-37ce-472b-ba0e-53cdaaa160ef.png)

**Comment by wo80** [comment]
If you want to support me, donations are very welcome. I do have bitcoin and zcash wallets. You should also have my private email address `christian.wo#######@tu-do######.de`, so feel free to contact me there.

**Comment by wo80** [comment]
Updated the gist again to include a post processing function
```c#
void PostProcess(IMesh mesh, params int[] constraints)
```
which returns
```
Processed 40 triangles in region connected to segment with label 4
Processed 39 triangles in region connected to segment with label 4
```

**Comment by epsi1on** [comment]
Thanks.
How to add some internal points (i mean the cyan points)?
![bitmap2](https://user-images.githubusercontent.com/7839141/153768502-b8463d24-26ed-4f45-bf25-c5f1cec58d3a.png)
Those point should be presented in triangulation result.

**Comment by wo80** [comment]
> How to add some internal points (i mean the cyan points)?

If you are using the code from the gist, just add the vertices in the `CreatePolygon()` method:
```c#
poly.Add(new Vertex(-3.5, 0.0, 5)); // 5 = vertex label
```
> I'm not able to send email to that mailbox. it gives me dns error

You'd have to replace the hash chars with correct values. If I remember correctly, we had email contact back in 2016, so you should have the address (I don't like to post my email address on public message boards).

Anyways, you can use [this contact form](http://wo80.bplaced.net/contact.html) with the following TAN: 51666

I've also written a little page with donation options: http://wo80.bplaced.net/donate.html

**Comment by epsi1on** [comment]
Thanks,
To preserve the quality, during mesh process some nodes are added to result. How can I find them?

**Comment by epsi1on** [comment]
Thanks.
Ho to identify the Traignles inside holes? i need to exclude them from result. And cannot use `RegionPointer`s to define regions. Because it is hard (for me) to programmatically find any point that located inside the a polygon.

**Comment by wo80** [comment]
I updated the gist once again. There are two approaches for finding holes, one using region pointers and one using a quadtree. Both should be pretty robust and efficient. Let know how those work out for you!

**Comment by epsi1on** [comment]
Thanks. My problem is this:
Because i'm getting the hole geometry as set of vertex from user. With you approach user must specify two items: hole vertex list and a sample point inside the hole's polygon (which you store inside `holes` list). the problem is second item (point inside the polygon) that should be automatically computed, and should not taken from user. I have two ways:

- Use another approach that do not need second item (point inside hole)
- Use a method to automatically compute the second item

I do not have any idea about walking in both ways! :)

**Comment by wo80** [comment]
That's a frequent problem when using Triangle and Triangle.NET's `Contour` class has a method which can help. It's a kinda heuristic approach, but it always worked for me:
```c#
var vertices = ... // The vertices that define the hole (have to be in order, obviously)
var contour = new Contour(vertices);
var p = contour.FindInteriorPoint(); // Point inside the contour.
```

**Comment by epsi1on** [comment]
> Do you mean the [Steiner points](https://www.iue.tuwien.ac.at/phd/fleischmann/node54.html) that are automatically added by Triangle for quality meshing?
> 
> Those will be added to the back of the mesh vertices list, so to find them (using Linq), just skip the input points
> 
> ```cs
>  mesh.Vertices.Skip(poly.Points.Count)
> ```

If a duplicated point is added to input then it is removed from mesh output, so i think `mesh.Vertices.Skip(poly.Points.Count)` is not a good general idea. I do work with ``Vertex.ID``, any ID that is not generated by me is new and generated by framework... Is it good you think?

> If you find the time, please review the code and let me know, whether you think this is a reasonable approach or not :-)
Is it ported from original C code?

**Comment by wo80** [comment]
> If a duplicated point is added to input then it is removed from mesh output, so i think `mesh.Vertices.Skip(poly.Points.Count)` is not a good general idea. I do work with `Vertex.ID`, any ID that is not generated by me is new and generated by framework... Is it good you think?

Duplicates will not be removed. They are marked as _undead_ and will simply not be part of the mesh topology. But they are still present in the vertices collection, see [TriangulatorTest.cs](https://github.com/wo80/Triangle.NET/blob/develop/src/Triangle.Tests/Meshing/Algorithm/TriangulatorTest.cs).

> Is it ported from original C code?

No, it's only available in the .NET version.
